dev/src/lobster/flat_gltf.h

```diff
@@ -0,0 +1,789 @@
+  _VERBOUT_ << "Size of Vertex:     " << sizeof(Vertex) << std::endl;
+
+  // perhaps an inconsistent use of data here but I'm not sure what to do
+  // sizeof returns a size_t which can be cast to anything since it's just a
+  // number
+  std::size_t pointer = 0;
+  glVertexAttribPointer(0, 4, GL_FLOAT, GL_FALSE, sizeof(Vertex),
+                        (const GLvoid *)pointer); // Position
+  pointer += sizeof(vertices[0].Position);
+  glVertexAttribPointer(1, 3, GL_FLOAT, GL_FALSE, sizeof(Vertex),
+                        (const GLvoid *)pointer); // Normal
+  pointer += sizeof(vertices[0].Normal);
+  glVertexAttribPointer(2, 2, GL_FLOAT, GL_FALSE, sizeof(Vertex),
+                        (const GLvoid *)pointer); // UV
+  pointer += sizeof(vertices[0].UV);
+  glVertexAttribPointer(3, 4, GL_FLOAT, GL_FALSE, sizeof(Vertex),
+                        (const GLvoid *)pointer); // Color (not used)
+  pointer += sizeof(vertices[0].Color);
+  glVertexAttribPointer(4, 4, GL_FLOAT, GL_FALSE, sizeof(Vertex),
+                        (const GLvoid *)pointer); // Weights
+  pointer += sizeof(vertices[0].Weights);
+  glVertexAttribPointer(5, 4, GL_FLOAT, GL_FALSE, sizeof(Vertex),
+                        (const GLvoid *)pointer); // Indices
+
+  // switch to the index buffer
+  glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, mesh.vao[2]);
+  // upload the indices, passing the size in bytes
+  glBufferData(GL_ELEMENT_ARRAY_BUFFER, indices.size() * sizeof(Index),
+               &indices[0], streaming ? GL_STREAM_DRAW : GL_STATIC_DRAW);
+  _VERBOUT_ << "Number of indices: " << indices.size() << std::endl;
+  _VERBOUT_ << "Size of Index:     " << sizeof(Index) << std::endl << std::endl;
+
+  // switch out of the VAO
+  glBindVertexArray(0);
+
+  // failure conditions are currently unimplemented
+  mesh.success = true;
+
+  // store the number of indices
+  mesh.indexCount = indices.size();
+
+  return mesh;
+}
+
+void DrawMesh(Mesh &mesh, float4x4 modelViewMat, float4x4 projMat) {
+  if (!mesh.success)
+    return;
+
+  float4x4 mvp = modelViewMat * projMat;
+
+  // the casting of each field means double4x4 could potentially be used instead
+  // TODO FEATURE: no model matrix means no normal matrix, but sacrifices must
+  // be made :/
+  const GLfloat glmats[16] = {
+      (GLfloat)mvp[0][0], (GLfloat)mvp[0][1], (GLfloat)mvp[0][2], (GLfloat)mvp[0][3],
+      (GLfloat)mvp[1][0], (GLfloat)mvp[1][1], (GLfloat)mvp[1][2], (GLfloat)mvp[1][3],
+      (GLfloat)mvp[2][0], (GLfloat)mvp[2][1], (GLfloat)mvp[2][2], (GLfloat)mvp[2][3],
+      (GLfloat)mvp[3][0], (GLfloat)mvp[3][1], (GLfloat)mvp[3][2], (GLfloat)mvp[3][3]
+  };
+
+  glUniformMatrix4fv(currentshader->mvp_i, 1, GL_FALSE, glmats);
+
+  // switch graphics memory to the VAO
+  glBindVertexArray(mesh.vao[0]);
+
+  glDrawElements(GL_TRIANGLES, mesh.indexCount, GL_UNSIGNED_INT,
+                 (const GLvoid *)0);
+}
+
+Texture LoadTexture(const GLvoid *data, GLsizei width, GLsizei height,
```

Comment:
If whoever is calling this could just call `CreateTexture` from `glinterface.h`, that would be better. Generally, the more types and functions you can use from there, the better. If those don't implement what you need, feel free to modify them.

Reply:
Yeah, that `CreateTexture` function looks pretty familiar. ^.^ As nothing is rendering at this point and I don't want to screw it up worse, I'd like to know a bit more about how to use it.

The parameters are self-explanatory, but let me just confirm: `dim` should be `int2(width * height).data()` and `tf` is a bitmask thing where `TF_NONE` is retro/blocky and some combination of flags gives you the standard trilinear filtering.

`TF_BUFFER_HAS_MIPS` freaks me out a bit because OpenGL generates its own damn mips. Not my responsibility when using a texture loader, right? My `LoadTexture` function is heavily automated and checks for what byte alignment things should get, how mipmapping and filtering should be handled, etc. That way I just give it the image buffer and it fills in the blanks. Surely `CreateTexture` must also have an automatic mode.

Reply:
`dim` is a pointer because it could either supply an `int2` or `int3`.

There are many ways to create mips, and sometimes it makes sense to create your own. Some compressed texture formats even include them. Not sure why that is scary. It is an option, if you look at the code you'll see it calls `glGenerateMipmap` if that flag is not specified.

Reply:
Yeah, I always `glGenerateMipmap` when possible. The main caveats are image buffers and npot textures on GLES2, but this code goes back to the past, back to the `.dds` days. I remember those days...